daemon/cmd/watchdogs.go

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/datapath/loader"
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/promise"
+)
+
+const epBPFProgWatchdog = "ep-bpf-prog-watchdog"
+
+type epBPFProgWatchdogConfig struct {
+	EndpointBPFProgWatchdogRate    time.Duration
+	EndpointBPFProgWatchdogEnabled bool
+}
+
+func (c epBPFProgWatchdogConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool("endpoint-bpf-prog-watchdog-enabled", c.EndpointBPFProgWatchdogEnabled,
+		"Enable endpoint BPF program watchdog")
+
+	flags.Duration("endpoint-bpf-prog-watchdog-rate", c.EndpointBPFProgWatchdogRate,
+		"Duration to trigger endpoint BPF programs load check watchdog")
+}
+
+type epBPFProgWatchdogParams struct {
+	cell.In
+
+	Config        epBPFProgWatchdogConfig
+	Logger        logrus.FieldLogger
+	Lifecycle     hive.Lifecycle
+	DaemonPromise promise.Promise[*Daemon]
+}
+
+var (
+	// EndpointBPFrogWatchdogCell triggers a job to ensure device tc programs remain loaded.
+	EndpointBPFrogWatchdogCell = cell.Module(
+		epBPFProgWatchdog,
+		"Periodically checks that endpoints BPF programs remain loaded",
+
+		cell.Config(epBPFProgWatchdogConfigDefault),
+		cell.Invoke(registerEndpointBPFProgWatchdog),
+	)
+
+	epBPFProgWatchdogConfigDefault = epBPFProgWatchdogConfig{
+		EndpointBPFProgWatchdogEnabled: true,
+		EndpointBPFProgWatchdogRate:    30 * time.Second,
+	}
+)
+
+func registerEndpointBPFProgWatchdog(p epBPFProgWatchdogParams) {
+	if !p.Config.EndpointBPFProgWatchdogEnabled {
+		return
+	}
+
+	var (
+		ctx, cancel = context.WithCancel(context.Background())
+		mgr         = controller.NewManager()
+	)
+	p.Lifecycle.Append(hive.Hook{
+		OnStart: func(hive.HookContext) error {
+			mgr.UpdateController(
+				epBPFProgWatchdog,
+				controller.ControllerParams{
+					Group: controller.NewGroup(epBPFProgWatchdog),
+					DoFunc: func(ctx context.Context) error {
+						d, err := p.DaemonPromise.Await(context.Background())
+						if err != nil {
+							return err
+						}
+						if d.ctx.Err() != nil {
+							mgr.RemoveAllAndWait()
+							return d.ctx.Err()
+						}
+						return checkEndpointBPFPrograms(ctx, d, p)
+					},
+					RunInterval: p.Config.EndpointBPFProgWatchdogRate,
+					Context:     ctx,
+				},
+			)
+
+			return nil
+		},
+		OnStop: func(hive.HookContext) error {
+			cancel()
+			mgr.RemoveAllAndWait()
```

Comment:
Same here regarding the remove all. I think we just want to remove this controller from the manager and cancel the context.